### ISSUE TITLE
RE-OPEN: Associating .ico files with ImageGlass 

### ISSUE THREAD

**Issue by daved163g** [d2phap/ImageGlass#1492]
### System information:
 - Windows OS version: 10
 - ImageGlass version:  8.7

### To Reproduce
1. Install IG
2. Set IG to be the default viewer or reset associations

### Actual behavior:
All filetypes that IG can handle have a new registry entry created for each one, e.g: HKEY_LOCAL_MACHINE\SOFTWARE\Classes\ImageGlass.AssocFile.JPG

This causes ALL other programs to use the IG icon when browsing or opening an image file. The IG icons are pretty ugly, so this behaviour is undesirable.

### Expected behavior:
It is OK for IG to use its own icons for itself. It is not ok for IG to force the use of its icons on all other programs.

### Screenshots:
Not needed - it's self-explanatory.

### Additional context:
This bug has been reopened because the previous one was closed without the problem being solved. Any solution must include a method for users to restore their systems to normal (i.e. to use the standard Windows default image icon where necessary, and to restore the default image icons that other programs have installed).

**Comment by d2phap** [comment]
> _All filetypes that IG can handle have a new registry entry created for each one, e.g: HKEY_LOCAL_MACHINE\SOFTWARE\Classes\ImageGlass.AssocFile.JPG_

This is the official way to set association in Windows, documented by Microsoft.



> _This causes ALL other programs to use the IG icon when browsing or opening an image file._ 

You can read these comments:
- https://github.com/d2phap/ImageGlass/issues/664#issuecomment-1018246003
- https://github.com/d2phap/ImageGlass/issues/664#issuecomment-1022305280
- https://github.com/d2phap/ImageGlass/issues/664#issuecomment-1196237285
- https://groups.google.com/g/imageglass/c/oPYXNnf3mrs/m/qNxmhB_MAQAJ

All them say the issue caused by Google Drive, not ImageGlass.


> The IG icons are pretty ugly, so this behaviour is undesirable.

The icons are beautiful which contributed by an IG user. If you don't like that, you can download other icon sets here: https://imageglass.org/themes, or delete all icons in Ext-Icons folder.

**Comment by daved163g** [comment]
Hi, thank you for responding so quickly. I do like ImageGlass in many ways and I thank you for your efforts. I have struggled to find a replacement for Windows Photos, which is awful and I replaced it with the old Windows Fax & Photo Viewer, but even that had its problems.

 

I had already read the posts you suggested. Google Drive is not the answer, nor is the wholesale removal of cached icons. These responses to the problem are typical of users trying anything they can think of to solve a problem that really should never have occurred in the first place. The internet is full of such misdirected advice. In any case, no user should be forced to dive deep into the operating system in order to resolve a problem.

 

Beauty is in the eye of the beholder.  For every person who loves them someone else will hate them!  I do a lot of photo editing and therefore use programs that load bunches of images on a constant basis, as their caches fill and they reload images.  A lot of the problem is actually the overbearing size of the large icons (see below where I remove the icons as you suggest).  I would have no objection to the ImageGlass icons being limited to the ImageGlass app, but they are not – they pervade everywhere except where other progams have done their own thing.  For example, I use the Faststone viewer and it chooses not to use the default Windows image icon (the blue mountains), but equally it does not force its method for unresolved images onto every other program. 

 

I’m sorry, but this is definitely an ImageGlass problem. Other programs just don’t act the same way. You might be following “a” recommended Windows way of doing things, but that is not always the best way and, to be clear, I can’t remember ever having come across an app that altered icons across the board. (I have been using Windows for over 30 years!)  

The Userchoice key is meant to override Windows default setting and that in itself should be a clue that perhaps this is not the correct mechanism to use because, clearly, it replaces a Windows default setting and therefore becomes active in all cases where other programs have not made their own arrangements. In other words, I think ImageGlass is using the wrong mechanism.

 

One of the issues with ImageGlass is that there seems to be no method to select individual image formats – it’s either all or nothing. Choosing “Reset…” or making ImageGlass the default viewer writes the registry keys for all image types. Choosing “Unregister” deletes all those keys. 

 

Thanks for a really innovative app. It is refreshing to come across something that isn’t just a clone of what’s gone before. I was particularly pleased that ImageGlass allows me to reconfigure the toolbar and to set the thumbnail bar to the right – really excellent options. I’m really glad you included colour management – any improvements in that direction are always welcome to photographers.

 

I tried your suggestion of deleting the icons from the ImageGlass folder (I’ve kept a copy for easy replacement, of course). This seems to have limited the use to smaller, more discrete icons in most cases.  It’s a reasonable compromise if not a proper fix and I still think the approach of using UserChoice is erroneous.

 

Having set the thumbnail bar to be vertical, I wonder if it would be possible to automatically set the up/down arrow keys to move the selection, as these keys are otherwise unused.

 

I’ve attached a picture of how I’ve set the toolbar, because it seems more logical to me and, importantly, reduces the risk of accidentally rotating an image if I miss the right arrow.

 

Thanks again,

Dave

**Comment by daved163g** [comment]
'Closed as not planned'. What does that mean?

This has been open since February and it's still a big problem. I read the comments as suggested and the reaction is mixed. It is actually only one thread with three entry points and it certainly does not give a definitive answer. Some people claim it is caused by Google Drive and others refute this. Personally I have never had Google Drive installed on my PC, so it cannot be the fault in my case.
I suggested earlier that some spurious registry keys are written which does not seem right to me because other programs don't do this - they use the existing keys for each file type to point to the program & icon to use.
Please do not confuse the .ico issue with the larger problem of ImageGlass icons taking over the whole system. They appear to be two different problems. No one should have to work as hard as some of the suggestioned in the comment above to resolve a problem. That is a sure indication of buggy software.

**Comment by daved163g** [comment]
Hi,

 

I have spent several days looking at how ImageGlass works and attach a couple of text files with my results. These are as much for my own benefit as anyone else – I usually make notes about issues regarding installations and bugs.

 

I am not a fan of the Windows 10 and 11 visual style (I use Open Shell to tweak the GUI to my taste) which explains various references to several “old” programs (I still use Windows 3.1 File Manager!). That said, I really do like the IG GUI, just not the icons everywhere! To be fair, a program should be able to use its own set of icons within itself, I just find it problematic that those same icons pervade the whole system. It is true that setting IG to be the default system viewer will cause its icons to be used system-wide, as you state BUT where a program has the facility to set file type associations on an individual basis – as does IG – then icons should not roll out across the whole system. This is where the fault lies, and is really mostly to do with the failure of point b) below.

 

My own approach to image-related software has always been to set my default image viewer to the old Windows Photos but as this is not colour managed it needed an update. The Windows 10 Photos app is colour managed but is otherwise absolutely awful. All other image software I always reject all associations, thus avoiding the problem of the icon changing globally.

 

I have found three main issues:

a)       Somewhat contrary to the statement below, which claims that each file type that IG will handle has its own registry entry, there are actually three entries for icons  – ARW, EPS and EXR – that contain a collection of some of the file types IG will handle, but not all. Why?

b)       The file type associations tab in Settings doesn’t work properly. To my knowledge it has never worked and the bug has been reported before. It this function worked properly it should at least be possible to reset things to the state they were before IG was installed. By setting individual file types (or all of them) in the setting should cause IG to open them by default but should not cause the IG icons to be used throughout the whole system. It is conventional to provide tick boxes for each type of file but IG has an all-or-nothing approach.

c)        In the forum where users raise an issue, it is common for some “solutions” to be unique to the particular user claiming a “fix”. This is certainly the case with the Google Drive fix because, never having installed or used Google Drive, that cannot possibly the case for me. It therefore becomes necessary to seek the true solution and, unfortunately, the references in the thread (and duplicated below) really have no bearing on the problem. Changing the icons set or even removing them as suggested did not resolve the problem and, again, this was noted somewhere in one of the threads.

 

Thanks for such a neat utility – it beats most others hands down, otherwise I wouldn’t have spent time with it!

 

Regards, Dave
